Implement SRP and auth functions

mmso opened the pull request.

mmso committed:
Implement SRP and auth functions

mmso committed:
Review unnecessary variables

mmso committed:
Mention pmcrypto in readme

wussler commented:
Using the code for the SRP tests it works reliably.

A possible modification for the future could be having the function `generateProofs` and `getSrp` return `sharedSession` as well, to allow for reuse.

twiss commented on README.md:
````diff
@@ -0,0 +1,104 @@
+};
+```
+
+where `ClientEphemal` and `ClientProof` should be sent in the auth request and the `ExpectedServerProof` should be validated against the received `ServerProof`.
````
```suggestion
where `ClientEphemeral` and `ClientProof` should be sent in the auth request and the `ExpectedServerProof` should be validated against the received `ServerProof`.
```

twiss commented on lib/constants.js:
```diff
@@ -0,0 +1,31 @@
+-----END PGP PUBLIC KEY BLOCK-----
+`;
+
+// Version 2 of bcrypt with 10 rounds.
```
```suggestion
// Version 2 of bcrypt with 2**10 rounds.
```
Just so that casual readers don't start writing blog posts about how insecure we are 🙂

twiss commented on lib/keys.js:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    // No salt, old-style
+    return password;
```
I know this is just copied from the previous code, but this seems scary to me. If we ever accidentally don't pass a salt for a new user, their password doesn't get hashed here, which is bad. Could we let this depend on the version perhaps, or at least when generating or upgrading a key, throw when there's no salt passed?

mmso replied:
Yes we can separate this

twiss commented on lib/keys.js:
```diff
@@ -0,0 +1,29 @@
+export const computeKeyPassword = async (password, salt) => {
+    if (salt && salt.length) {
+        const saltBinary = binaryStringToArray(decodeBase64(salt));
+        const hash = await bcrypt.hash(password, BCRYPT_PREFIX + bcrypt.encodeBase64(saltBinary, 16));
```
It's probably insignificant compared to the time bcrypt takes, but we're base64 decoding and encoding here just to change the alphabet from  
`ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/` to `./ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789`.
If ever needed, it's possible to optimize this by just replacing all of the characters by the one directly below it. Probably not needed now though.

twiss commented on test/passwords.data.js:
```diff
@@ -0,0 +1,1035 @@
+    253,
+    40,
+    41
+]);
```
Could you format these on one line perhaps so that it's a bit easier to read?

twiss commented on test/srp.data.js:
```diff
@@ -0,0 +1,295 @@
+    152,
+    98,
+    175
+];
```
And this one

twiss commented on lib/passwords.js:
```diff
@@ -0,0 +1,110 @@
+        SHA512(binaryStringToArray(str + '\x00')),
+        SHA512(binaryStringToArray(str + '\x01')),
+        SHA512(binaryStringToArray(str + '\x02')),
+        SHA512(binaryStringToArray(str + '\x03'))
```
This might again be premature optimization, but both callers of this function call `arrayToBinaryString` before passing `str` here, and we're converting it back here 4 times. It might be easier to just pass it as a Uint8Array. You might even get away with reusing only one, even though the functions are async, like
```js
const list = [];
const arr = concatArrays([input, new Uint8Array([0])]);
for (let i = 1; i <= 4; i++) {
    list.push(SHA512(arr));
    arr[arr.length - 1] = i;
}
return concatArrays(await Promise.all(list));
```

twiss commented on lib/utils/modulus.js:
```diff
@@ -0,0 +1,46 @@
+ * Get modulus keys.
+ * @return {Promise}
+ */
+export const getModulusKeys = () => readArmoredWithCache(SRP_MODULUS_KEY);
```
Rather than json encoding the modulus key and constructing a cache based on that, it might be easier to just specifically add a variable to cache the return value of this function here, since the cache is not used for anything else AFAICT?

twiss commented on lib/srp.js:
```diff
@@ -0,0 +1,273 @@
+import { BigNumber, Modulus } from 'asmcrypto.js/asmcrypto.all.es8';
```
```suggestion
import { BigNumber, Modulus } from 'asmcrypto.js/dist_e8/bignum/bignum';
```
Not sure how well tree shaking works with the `all.es8` version but since they export a separate file anyway it might be good to use that.
Also, openpgp.js indirectly imports this file as well (but doesn't export it). At some point in the future we should probably include openpgpjs directly in pmcrypto, so that we can deduplicate that, have two builds of pmcrypto (compat and non compat), and then we also don't need to wait for `init(openpgp)` here anymore.

twiss commented on lib/utils/bigNumber.js:
```diff
@@ -0,0 +1,29 @@
+import { BigNumber } from 'asmcrypto.js/asmcrypto.all.es8';
```
```suggestion
import { BigNumber } from 'asmcrypto.js/dist_e8/bignum/bignum';
```

twiss commented on lib/srp.js:
```diff
@@ -0,0 +1,273 @@
+    );
+
+    return {
+        ClientEphemeral: fromBN(len, clientEphemeral),
```
Again I know this is copy+pasted but `fromBN(len, clientEphemeral)` is called three times here, and `fromBN(len, sharedSession)` twice. Might want to pull them out into variables.

twiss commented on lib/srp.js:
```diff
@@ -0,0 +1,273 @@
+    const sharedSession = modulus.power(subtracted, exponent);
+
+    const clientProof = await srpHasher(
+        concatArrays([fromBN(len, clientEphemeral), fromBN(len, serverEphemeral), fromBN(len, sharedSession)])
```
And we already have `serverEphemeralArray` here

twiss commented on lib/passwords.js:
```diff
@@ -0,0 +1,110 @@
+ */
+const hashPassword1 = async (password, username, modulus) => {
+    const value = binaryStringToArray(encodeUtf8(username.toLowerCase()));
+    // eslint-disable-next-line new-cap
```
```suggestion
```
More nits: this shouldn't be needed for `unsafeMD5`, although you might need it for the `SHA512` calls in various places?

twiss commented on lib/constants.js:
```diff
@@ -0,0 +1,31 @@
+`;
+
+// Version 2 of bcrypt with 10 rounds.
+// https://passlib.readthedocs.io/en/stable/modular_crypt_format.html
```
```suggestion
// https://en.wikipedia.org/wiki/Bcrypt#Description
```
A link that backs that up

twiss reviewed (commented):
Mostly nitpicks below 🙂

twiss committed:
Update lib/constants.js

Co-Authored-By: mmso <mmso@protonmail.com>

twiss committed:
Update README.md

Co-Authored-By: mmso <mmso@protonmail.com>

twiss committed:
Update lib/constants.js

Co-Authored-By: mmso <mmso@protonmail.com>

mmso committed:
Require salt and password for compute key password

twiss committed:
Update lib/passwords.js

Co-Authored-By: mmso <mmso@protonmail.com>

twiss committed:
'asmcrypto.js/dist_es8/bignum/bignum'

Co-Authored-By: mmso <mmso@protonmail.com>

twiss commented on lib/srp.js:
```diff
@@ -0,0 +1,273 @@
+import { BigNumber, Modulus } from 'asmcrypto.js/dist_e8/bignum/bignum';
```
```suggestion
import { BigNumber, Modulus } from 'asmcrypto.js/dist_es8/bignum/bignum';
```
Sorry, I made a typo here :S Don't forget to test my suggestions before merging this PR because I clearly didn't  😅

twiss commented on lib/utils/bigNumber.js:
```diff
@@ -0,0 +1,29 @@
+import { BigNumber } from 'asmcrypto.js/dist_e8/bignum/bignum';
```
```suggestion
import { BigNumber } from 'asmcrypto.js/dist_es8/bignum/bignum';
```

mmso committed:
Pull into variables

mmso committed:
Fix newlines

mmso committed:
Change modulus key caching

mmso committed:
Camel case instead of pascal case

Also add sharedSession returned from getSrp and remove ModulusID
returned from getRandomSrpVerifier

mmso committed:
Review with @twiss

mmso committed:
Change getAuthVersion signature
